rust/agama-lib/src/storage/client/dasd.rs

```diff
@@ -66,6 +76,149 @@ impl<'a> DASDClient<'a> {
+        Ok(DASDConfig::default())
+    }
+
+    pub async fn set_config(&self, config: DASDConfig) -> Result<(), ServiceError> {
+        // at first probe to ensure we work on real system info
+        self.probe().await?;
+        self.config_activate(&config).await?;
+        self.config_format(&config).await?;
+        self.config_set_diag(&config).await?;
+
+        Ok(())
+    }
+
+    async fn config_activate(&self, config: &DASDConfig) -> Result<(), ServiceError> {
+        let pairs = self.config_pairs(config).await?;
+        let to_activate: Vec<&str> = pairs
+            .iter()
+            .filter(|(system, _config)| system.enabled == false)
+            .filter(|(_system, config)| {
+                config.state.clone().unwrap_or_default() == DASDDeviceState::Active
+            })
+            .map(|(system, _config)| system.id.as_str())
+            .collect();
+        self.enable(&to_activate).await?;
+
+        let pairs = self.config_pairs(config).await?;
+        let to_deactivate: Vec<&str> = pairs
+            .iter()
+            .filter(|(system, _config)| system.enabled == true)
+            .filter(|(_system, config)| config.state == Some(DASDDeviceState::Offline))
+            .map(|(system, _config)| system.id.as_str())
+            .collect();
+        self.disable(&to_deactivate).await?;
+
+        if !to_activate.is_empty() || !to_deactivate.is_empty() {
+            // reprobe after calling enable. TODO: check if it is needed or callbacks take into action and update it automatically
+            self.probe().await?;
+        }
+        Ok(())
+    }
+
+    async fn config_format(&self, config: &DASDConfig) -> Result<(), ServiceError> {
+        let pairs = self.config_pairs(config).await?;
+        let to_format: Vec<&str> = pairs
+            .iter()
+            .filter(|(system, config)| {
+                if config.format == Some(true) {
+                    true
+                } else if config.format == None {
+                    !system.formatted
+                } else {
+                    false
+                }
+            })
+            .map(|(system, _config)| system.id.as_str())
+            .collect();
+
+        if !to_format.is_empty() {
+            let job_path = self.format(&to_format).await?;
+            self.wait_for_format(job_path).await?;
+            // reprobe after calling format. TODO: check if it is needed or callbacks take into action and update it automatically
+            // also do we need to wait here for finish of format progress?
+            self.probe().await?;
+        }
+        Ok(())
+    }
+
+    async fn wait_for_format(&self, job_path: String) -> Result<(), ServiceError> {
+        let mut finished = false;
+        let jobs_client = JobsClient::new(
+            self.connection.clone(),
+            "org.opensuse.Agama.Storage1",
+            "/org/opensuse/Agama/Storage1/jobs",
+        )
+        .await?;
+        while !finished {
```

Comment:
I guess this one works just fine, but I wonder whether it could be possible to rely on signals (I do not know if jobs emit signals).